File: src/artemis_calendar/web/static/js/pages/blend.js
/**
 * Vote Simulator — configure attribute-biased voting blocks,
 * run in-memory simulation, compare per-block top-13 vs blended result,
 * and inspect individual sample ballots to see how bias shapes selections.
 */

import { showImageDetail } from '../components/image-detail.js';

const SCENARIO_PRESETS = {
  balanced:       { label: 'Balanced (all 10)',  values: { earth_moon: 10, earth_only: 10, moon_only: 10, sun_only: 10, crescent: 10, eclipse: 10, crew: 10, equipment: 10, moon_sun: 10, neutral: 10 } },
  celestial:      { label: 'Celestial Focus',    values: { earth_moon: 20, earth_only: 15, moon_only: 20, sun_only: 10, crescent: 20, eclipse: 10, crew: 0, equipment: 0, moon_sun: 5, neutral: 10 } },
  human:          { label: 'Human Interest',     values: { earth_moon: 0, earth_only: 5, moon_only: 0, sun_only: 0, crescent: 0, eclipse: 0, crew: 40, equipment: 40, moon_sun: 0, neutral: 15 } },
  dramatic:       { label: 'Dramatic Shots',     values: { earth_moon: 15, earth_only: 0, moon_only: 0, sun_only: 10, crescent: 30, eclipse: 25, crew: 0, equipment: 0, moon_sun: 10, neutral: 10 } },
  neutral_only:   { label: 'Neutral Only',       values: { earth_moon: 0, earth_only: 0, moon_only: 0, sun_only: 0, crescent: 0, eclipse: 0, crew: 0, equipment: 0, moon_sun: 0, neutral: 100 } },
};

// Embedded presets — renders without API so GitHub Pages shows the UI
const DEFAULT_BLOCKS = [
  { block_id: 'earth_moon', label: 'Earth+Moon Fans', description: 'Prefer images showing both Earth and Moon together', voter_count: 10, votes_per_voter: 10, preference_weight: 3.0, randomness_weight: 0.5, preference_rules: { all_of: ['earth', 'moon'], any_of: [], none_of: [] } },
  { block_id: 'earth_only', label: 'Earth Only', description: 'Prefer Earth views without Moon or Sun', voter_count: 10, votes_per_voter: 10, preference_weight: 3.0, randomness_weight: 0.5, preference_rules: { all_of: ['earth'], any_of: [], none_of: ['moon', 'sun'] } },
  { block_id: 'moon_only', label: 'Moon Only', description: 'Prefer Moon views without Earth or Sun', voter_count: 10, votes_per_voter: 10, preference_weight: 3.0, randomness_weight: 0.5, preference_rules: { all_of: ['moon'], any_of: [], none_of: ['earth', 'sun'] } },
  { block_id: 'sun_only', label: 'Sun Only', description: 'Prefer images featuring the Sun without Earth or Moon', voter_count: 10, votes_per_voter: 10, preference_weight: 3.0, randomness_weight: 0.5, preference_rules: { all_of: ['sun'], any_of: [], none_of: ['earth', 'moon'] } },
  { block_id: 'crescent', label: 'Crescent Fans', description: 'Prefer crescent shapes — crescent Moon or crescent Earth', voter_count: 10, votes_per_voter: 10, preference_weight: 3.0, randomness_weight: 0.5, preference_rules: { all_of: [], any_of: ['crescent_moon', 'crescent_earth'], none_of: [] } },
  { block_id: 'eclipse', label: 'Eclipse / Sun+Moon', description: 'Prefer images with Sun and Moon together, or Sun with lens flare', voter_count: 10, votes_per_voter: 10, preference_weight: 3.0, randomness_weight: 0.5, preference_rules: { all_of: [], any_of: ['sun', 'lens_flare'], none_of: [] } },
  { block_id: 'crew', label: 'Crew / People', description: 'Prefer images showing astronauts, crew, hands, or selfies', voter_count: 10, votes_per_voter: 10, preference_weight: 3.0, randomness_weight: 0.5, preference_rules: { all_of: [], any_of: ['astronaut', 'crew', 'hand', 'selfie'], none_of: [] } },
  { block_id: 'equipment', label: 'Equipment / Hardware', description: 'Prefer images of spacecraft, vehicles, habitat, or interior views', voter_count: 10, votes_per_voter: 10, preference_weight: 3.0, randomness_weight: 0.5, preference_rules: { all_of: [], any_of: ['spacecraft', 'vehicle', 'habitat', 'interior', 'porthole'], none_of: [] } },
  { block_id: 'moon_sun', label: 'Moon+Sun Fans', description: 'Prefer images showing Moon and Sun together', voter_count: 10, votes_per_voter: 10, preference_weight: 3.0, randomness_weight: 0.5, preference_rules: { all_of: ['moon', 'sun'], any_of: [], none_of: [] } },
  { block_id: 'neutral', label: 'Neutral', description: 'No attribute preference — votes by general appeal + noise', voter_count: 10, votes_per_voter: 10, preference_weight: 0.0, randomness_weight: 1.0, preference_rules: { all_of: [], any_of: [], none_of: [] } },
];

let presets = null;
let lastResult = null;

export async function render(el) {
  // Try API first, fall back to embedded defaults
  presets = await fetch('/api/blend/presets').then(r => {
    if (!r.ok) throw new Error(r.status);
    return r.json();
  }).catch(() => ({ blocks: DEFAULT_BLOCKS, attributes: [] }));

  el.innerHTML = `
    <div class="page-header">
      <h1>Vote Simulator</h1>
      <div class="controls">
        <label>Scenario
          <select class="scenario-select">
            <option value="">Custom</option>
            ${Object.entries(SCENARIO_PRESETS).map(([k, v]) =>
              `<option value="${k}">${v.label}</option>`
            ).join('')}
          </select>
        </label>
        <label>Seed
          <input type="number" class="seed-input" value="42" min="1" max="99999" style="width:70px">
        </label>
        <button class="btn run-btn">Run Simulation</button>
      </div>
    </div>

    <details class="blend-explainer" open>
      <summary>How this works</summary>
      <p>Each <strong>voter block</strong> represents a group of voters who prefer certain image attributes.
      When you run a simulation, each voter is shown 50 random images and picks their top 5.
      Biased voters score images higher when they match the block's attribute rules —
      so "Earth+Moon Fans" will disproportionately select images tagged with both <code>earth</code> and <code>moon</code>.</p>
      <p>The <strong>blended result</strong> combines all voters' picks. Adjust the voter counts to see how
      different electorate compositions change which images rise to the top.</p>
      <p>Expand <strong>Sample Ballots</strong> below the results to see exactly what individual voters saw and chose.</p>
    </details>

    <div class="blend-config" id="blend-config"></div>
    <div class="blend-summary" id="blend-summary"></div>
    <div class="blend-results" id="blend-results"></div>
    <div class="blend-ballots" id="blend-ballots"></div>
  `;

  renderBlockCards(el);
  bindEvents(el);
}

function renderBlockCards(el) {
  const container = el.querySelector('#blend-config');
  container.innerHTML = presets.blocks.map(b => `
    <div class="blend-block-card" data-block="${b.block_id}">
      <div class="blend-block-header">
        <strong>${b.label}</strong>
        <span class="blend-block-desc">${b.description}</span>
      </div>
      <div class="blend-block-rules">
        ${formatRules(b.preference_rules)}
      </div>
      <div class="blend-slider-row">
        <label>Voters: <span class="voter-val">${b.voter_count}</span></label>
        <input type="range" class="voter-slider" data-block="${b.block_id}"
               min="0" max="100" value="${b.voter_count}" step="5">
      </div>
      <div class="blend-weights">
        <span>Pref weight: ${b.preference_weight}</span>
        <span>Noise: ${b.randomness_weight}</span>
      </div>
    </div>
  `).join('');
}

function formatRules(rules) {
  const parts = [];
  if (rules.all_of?.length) parts.push(`<span class="rule-tag rule-all">ALL: ${rules.all_of.join(', ')}</span>`);
  if (rules.any_of?.length) parts.push(`<span class="rule-tag rule-any">ANY: ${rules.any_of.join(', ')}</span>`);
  if (rules.none_of?.length) parts.push(`<span class="rule-tag rule-none">NONE: ${rules.none_of.join(', ')}</span>`);
  return parts.length ? parts.join(' ') : '<span class="rule-tag rule-neutral">No attribute bias</span>';
}

function bindEvents(el) {
  el.querySelectorAll('.voter-slider').forEach(slider => {
    slider.addEventListener('input', () => {
      slider.closest('.blend-block-card').querySelector('.voter-val').textContent = slider.value;
      el.querySelector('.scenario-select').value = '';
    });
  });

  el.querySelector('.scenario-select').addEventListener('change', (e) => {
    const key = e.target.value;
    if (!key) return;
    const preset = SCENARIO_PRESETS[key];
    Object.entries(preset.values).forEach(([blockId, count]) => {
      const slider = el.querySelector(`.voter-slider[data-block="${blockId}"]`);
      if (slider) {
        slider.value = count;
        slider.closest('.blend-block-card').querySelector('.voter-val').textContent = count;
      }
    });
  });

  el.querySelector('.run-btn').addEventListener('click', () => runSimulation(el));
}

async function runSimulation(el) {
  const btn = el.querySelector('.run-btn');
  const resultsEl = el.querySelector('#blend-results');
  const summaryEl = el.querySelector('#blend-summary');
  const ballotsEl = el.querySelector('#blend-ballots');

  const blocks = presets.blocks.map(b => {
    const slider = el.querySelector(`.voter-slider[data-block="${b.block_id}"]`);
    return {
      block_id: b.block_id,
      label: b.label,
      voter_count: parseInt(slider?.value || '0'),
      votes_per_voter: b.votes_per_voter,
      preference_weight: b.preference_weight,
      randomness_weight: b.randomness_weight,
      preference_rules: b.preference_rules,
    };
  });

  const seed = parseInt(el.querySelector('.seed-input').value) || 42;

  btn.disabled = true;
  btn.textContent = 'Simulating...';
  resultsEl.innerHTML = '<div class="loading">Running simulation...</div>';
  summaryEl.innerHTML = '';
  ballotsEl.innerHTML = '';

  try {
    const resp = await fetch('/api/blend/simulate', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ blocks, seed }),
    });

    if (!resp.ok) {
      const err = await resp.json().catch(() => ({ detail: resp.statusText }));
      throw new Error(err.detail || resp.statusText);
    }

    lastResult = await resp.json();
    renderSummary(summaryEl, lastResult);
    renderResults(resultsEl, lastResult);
    renderSampleBallots(ballotsEl, lastResult);
  } catch (err) {
    const isOffline = err.message.includes('Failed to fetch') || err.message.includes('404') || err.message.includes('Internal Server Error');
    resultsEl.innerHTML = isOffline
      ? `<div class="blend-offline-msg">
           <strong>Simulation requires the local server</strong>
           <p>The vote simulator runs in-memory on the backend and is only available when the Artemis server is running locally on port 8070.</p>
           <p>The block configuration UI above shows the available voter types and their attribute bias rules.</p>
         </div>`
      : `<p class="error">Simulation failed: ${err.message}</p>`;
  } finally {
    btn.disabled = false;
    btn.textContent = 'Run Simulation';
  }
}

function renderSummary(el, data) {
  const activeBlocks = data.blocks.filter(b => b.voter_count > 0);
  el.innerHTML = `
    <div class="blend-summary-bar">
      <span><strong>${data.total_voters}</strong> voters</span>
      <span><strong>${data.total_ballots}</strong> ballots</span>
      <span><strong>${data.eligible_images?.toLocaleString() || '?'}</strong> eligible images</span>
      <span><strong>${activeBlocks.length}</strong> active blocks</span>
      <span>Seed: <strong>${data.seed}</strong></span>
      <span title="Min CLIP confidence for attribute labels">Conf &ge; ${(data.min_confidence || 0.9).toFixed(2)}</span>
      <span title="Exclude images with this % or more dark pixels">Dark &lt; ${((data.max_dark_ratio || 0.92) * 100).toFixed(0)}%</span>
    </div>
  `;
}

function renderResults(el, data) {
  const blendedSet = new Set(data.blended.map(img => img.image_sk));

  const imageCounts = {};
  const allColumns = [...data.blocks, { block_id: '_blended', top_13: data.blended }];
  for (const col of allColumns) {
    for (const img of col.top_13) {
      imageCounts[img.image_sk] = (imageCounts[img.image_sk] || 0) + 1;
    }
  }

  const columns = data.blocks.map(b =>
    renderColumn(b.label, `${b.voter_count} voters, ${b.ballot_count} ballots`, b.top_13, blendedSet, imageCounts)
  );
  columns.push(
    renderColumn('Blended Result', `${data.total_voters} voters combined`, data.blended, blendedSet, imageCounts, true)
  );

  el.innerHTML = `
    <h2 class="blend-section-title">Top 13 Per Block vs Blended</h2>
    <div class="blend-columns">${columns.join('')}</div>
  `;

  el.querySelectorAll('.blend-thumb').forEach(thumb => {
    thumb.addEventListener('click', () => showImageDetail(parseInt(thumb.dataset.sk)));
  });
}

function renderColumn(title, subtitle, images, blendedSet, imageCounts, isBlended = false) {
  const items = images.map((img, i) => {
    const inBlended = blendedSet.has(img.image_sk);
    const count = imageCounts[img.image_sk] || 1;
    const overlapClass = count > 1 ? `blend-overlap-${Math.min(count, 4)}` : '';
    const blendedMark = !isBlended && inBlended ? ' blend-in-blended' : '';
    const rate = (img.selection_rate * 100).toFixed(1);

    return `
      <div class="blend-thumb ${overlapClass}${blendedMark}" data-sk="${img.image_sk}" title="SK ${img.image_sk} — selected ${img.selection_count}/${img.shown_count} times shown">
        <span class="blend-rank">${i + 1}</span>
        <img src="/thumbs/${img.nasa_id}.jpg" alt="${img.nasa_id}" loading="lazy">
        <div class="blend-rate-bar">
          <div class="blend-rate-fill" style="width:${Math.min(rate, 100)}%"></div>
        </div>
        <span class="blend-rate-label">${rate}%</span>
      </div>
    `;
  }).join('');

  return `
    <div class="blend-column ${isBlended ? 'blend-column-blended' : ''}">
      <div class="blend-column-header">
        <strong>${title}</strong>
        <span>${subtitle}</span>
      </div>
      ${items}
    </div>
  `;
}

// ---------------------------------------------------------------------------
// Sample Ballots — show representative voter records
// ---------------------------------------------------------------------------

function renderSampleBallots(el, data) {
  const allBallots = data.blocks.flatMap(b => b.sample_ballots || []);
  if (!allBallots.length) {
    el.innerHTML = '';
    return;
  }

  const blockLabels = {};
  for (const b of data.blocks) blockLabels[b.block_id] = b.label;

  const accordions = allBallots.map(ballot => {
    const selected = ballot.images.filter(img => img.was_selected);
    const notSelected = ballot.images.filter(img => !img.was_selected);

    return `
      <details class="ballot-detail">
        <summary class="ballot-summary">
          <strong>${blockLabels[ballot.block_id] || ballot.block_id}</strong>
          — Voter #${ballot.voter_number}, Ballot #${ballot.ballot_number}
          <span class="ballot-meta">Shown ${ballot.images.length} images, selected ${selected.length}</span>
        </summary>
        <div class="ballot-body">
          <div class="ballot-section">
            <h4 class="ballot-section-title">Selected (top ${selected.length} by utility score)</h4>
            <p class="ballot-section-desc">These images scored highest for this voter. The utility combines base appeal, attribute match, and random noise.</p>
            <div class="ballot-grid ballot-grid-selected">
              ${selected.map(img => renderBallotImage(img, true)).join('')}
            </div>
          </div>
          <div class="ballot-section">
            <h4 class="ballot-section-title">Not Selected (${notSelected.length} remaining)</h4>
            <p class="ballot-section-desc">These images were shown but scored lower. Notice the attribute match scores — biased voters systematically rank non-matching images lower.</p>
            <div class="ballot-grid ballot-grid-rejected">
              ${notSelected.map(img => renderBallotImage(img, false)).join('')}
            </div>
          </div>
        </div>
      </details>
    `;
  }).join('');

  el.innerHTML = `
    <h2 class="blend-section-title">Sample Ballots</h2>
    <p class="blend-section-desc">Each ballot below is one voter's actual decision: 50 random images shown, 5 selected. Expand to see <em>why</em> each image was or wasn't chosen.</p>
    ${accordions}
  `;

  el.querySelectorAll('.ballot-img').forEach(img => {
    img.addEventListener('click', () => showImageDetail(parseInt(img.dataset.sk)));
  });
}

function renderBallotImage(img, isSelected) {
  const matchBadges = img.matched_attrs.length
    ? img.matched_attrs.map(a => `<span class="ballot-attr-tag">${a}</span>`).join('')
    : '<span class="ballot-attr-none">no match</span>';

  return `
    <div class="ballot-img ${isSelected ? 'ballot-img-selected' : ''}" data-sk="${img.image_sk}">
      <img src="/thumbs/${img.nasa_id}.jpg" alt="${img.nasa_id}" loading="lazy">
      <div class="ballot-img-scores">
        <div class="ballot-score-row">
          <span class="ballot-score-label">Utility</span>
          <span class="ballot-score-val">${img.utility.toFixed(3)}</span>
        </div>
        <div class="ballot-score-row">
          <span class="ballot-score-label">Attr match</span>
          <span class="ballot-score-val">${img.attribute_match.toFixed(2)}</span>
        </div>
        <div class="ballot-score-row">
          <span class="ballot-score-label">Base appeal</span>
          <span class="ballot-score-val">${img.base_appeal.toFixed(3)}</span>
        </div>
        <div class="ballot-attrs">${matchBadges}</div>
      </div>
    </div>
  `;
}

File: src/artemis_calendar/web/routes/blend.py
"""Voter blend simulator — in-memory synthetic voting with attribute-biased blocks."""

from __future__ import annotations

import hashlib
import random
from collections import defaultdict

import duckdb
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from artemis_calendar.synthetic.block_generator import (
    BATCH_SELECT_COUNT,
    BATCH_SHOW_COUNT,
    compute_attribute_match,
    compute_block_utility,
)
from artemis_calendar.vision.voting_config import PreferenceRule, VotingBlockConfig
from artemis_calendar.web.db import get_db

router = APIRouter(prefix="/api/blend", tags=["blend"])

SAMPLES_PER_BLOCK = 3  # Number of example ballots to capture per block
DEFAULT_MIN_CONFIDENCE = 0.90  # Higher than the global 0.80 acceptance threshold
DEFAULT_MAX_DARK_RATIO = 0.92  # Exclude images that are >= 92% dark pixels


def _load_simulation_images(
    conn: duckdb.DuckDBPyConnection,
    min_confidence: float = DEFAULT_MIN_CONFIDENCE,
    max_dark_ratio: float = DEFAULT_MAX_DARK_RATIO,
) -> tuple[dict[int, set[str]], dict[int, str]]:
    """Load images eligible for simulation with quality-filtered attributes.

    Filters:
    - Vote-pool images only
    - Excludes non-master dedup members (near-duplicates)
    - Excludes images that are >= max_dark_ratio dark pixels
    - Only includes attributes with confidence >= min_confidence

    Returns (image_attrs, sid_map) where:
    - image_attrs: {image_sk: {attribute_codes}}
    - sid_map: {image_sk: source_image_id}
    """
    # Get eligible image SKs: vote-pool, not suppressed duplicate, not too dark
    eligible_rows = conn.execute(
        """
        SELECT di.image_sk, di.source_image_id
        FROM dim_image di
        LEFT JOIN feature_image_visual v ON v.image_sk = di.image_sk
        WHERE di.vote_pool_flag = true
          AND di.image_sk NOT IN (
              SELECT image_sk FROM dedup_image_member WHERE is_master = false
          )
          AND COALESCE(v.dark_pixel_ratio, 0) < ?
        """,
        [max_dark_ratio],
    ).fetchall()

    eligible_sks = {int(r[0]) for r in eligible_rows}
    sid_map = {int(r[0]): r[1] for r in eligible_rows}

    # Load attributes only for eligible images at higher confidence
    attr_rows = conn.execute(
        """
        SELECT fa.image_sk, fa.attribute_code
        FROM feature_image_attribute fa
        WHERE fa.image_sk IN (SELECT UNNEST(?::BIGINT[]))
          AND fa.confidence_score >= ?
        """,
        [list(eligible_sks), min_confidence],
    ).fetchall()

    image_attrs: dict[int, set[str]] = {sk: set() for sk in eligible_sks}
    for image_sk, code in attr_rows:
        image_attrs[int(image_sk)].add(code)

    return image_attrs, sid_map

# ---------------------------------------------------------------------------
# Preset block definitions
# ---------------------------------------------------------------------------

PRESET_BLOCKS = [
    {
        "block_id": "earth_moon",
        "label": "Earth+Moon Fans",
        "description": "Prefer images showing both Earth and Moon together",
        "voter_count": 10,
        "votes_per_voter": 10,
        "preference_weight": 3.0,
        "randomness_weight": 0.5,
        "preference_rules": {"all_of": ["earth", "moon"], "any_of": [], "none_of": []},
    },
    {
        "block_id": "earth_only",
        "label": "Earth Only",
        "description": "Prefer Earth views without Moon or Sun",
        "voter_count": 10,
        "votes_per_voter": 10,
        "preference_weight": 3.0,
        "randomness_weight": 0.5,
        "preference_rules": {
            "all_of": ["earth"],
            "any_of": [],
            "none_of": ["moon", "sun"],
        },
    },
    {
        "block_id": "moon_only",
        "label": "Moon Only",
        "description": "Prefer Moon views without Earth or Sun",
        "voter_count": 10,
        "votes_per_voter": 10,
        "preference_weight": 3.0,
        "randomness_weight": 0.5,
        "preference_rules": {
            "all_of": ["moon"],
            "any_of": [],
            "none_of": ["earth", "sun"],
        },
    },
    {
        "block_id": "sun_only",
        "label": "Sun Only",
        "description": "Prefer images featuring the Sun without Earth or Moon",
        "voter_count": 10,
        "votes_per_voter": 10,
        "preference_weight": 3.0,
        "randomness_weight": 0.5,
        "preference_rules": {
            "all_of": ["sun"],
            "any_of": [],
            "none_of": ["earth", "moon"],
        },
    },
    {
        "block_id": "crescent",
        "label": "Crescent Fans",
        "description": "Prefer crescent shapes — crescent Moon or crescent Earth",
        "voter_count": 10,
        "votes_per_voter": 10,
        "preference_weight": 3.0,
        "randomness_weight": 0.5,
        "preference_rules": {
            "all_of": [],
            "any_of": ["crescent_moon", "crescent_earth"],
            "none_of": [],
        },
    },
    {
        "block_id": "eclipse",
        "label": "Eclipse / Sun+Moon",
        "description": "Prefer images with Sun and Moon together, or Sun with lens flare",
        "voter_count": 10,
        "votes_per_voter": 10,
        "preference_weight": 3.0,
        "randomness_weight": 0.5,
        "preference_rules": {
            "all_of": [],
            "any_of": ["sun", "lens_flare"],
            "none_of": [],
        },
    },
    {
        "block_id": "crew",
        "label": "Crew / People",
        "description": "Prefer images showing astronauts, crew, hands, or selfies",
        "voter_count": 10,
        "votes_per_voter": 10,
        "preference_weight": 3.0,
        "randomness_weight": 0.5,
        "preference_rules": {
            "all_of": [],
            "any_of": ["astronaut", "crew", "hand", "selfie"],
            "none_of": [],
        },
    },
    {
        "block_id": "equipment",
        "label": "Equipment / Hardware",
        "description": "Prefer images of spacecraft, vehicles, habitat, or interior views",
        "voter_count": 10,
        "votes_per_voter": 10,
        "preference_weight": 3.0,
        "randomness_weight": 0.5,
        "preference_rules": {
            "all_of": [],
            "any_of": ["spacecraft", "vehicle", "habitat", "interior", "porthole"],
            "none_of": [],
        },
    },
    {
        "block_id": "moon_sun",
        "label": "Moon+Sun Fans",
        "description": "Prefer images showing Moon and Sun together",
        "voter_count": 10,
        "votes_per_voter": 10,
        "preference_weight": 3.0,
        "randomness_weight": 0.5,
        "preference_rules": {"all_of": ["moon", "sun"], "any_of": [], "none_of": []},
    },
    {
        "block_id": "neutral",
        "label": "Neutral",
        "description": "No attribute preference — votes by general appeal + noise",
        "voter_count": 10,
        "votes_per_voter": 10,
        "preference_weight": 0.0,
        "randomness_weight": 1.0,
        "preference_rules": {"all_of": [], "any_of": [], "none_of": []},
    },
]


# ---------------------------------------------------------------------------
# Request / response models
# ---------------------------------------------------------------------------


class BlockInput(BaseModel):
    block_id: str
    label: str = ""
    voter_count: int = Field(ge=0, le=200)
    votes_per_voter: int = Field(default=10, ge=1, le=50)
    preference_weight: float = Field(default=3.0, ge=0.0, le=10.0)
    randomness_weight: float = Field(default=0.5, ge=0.0, le=5.0)
    preference_rules: dict = Field(
        default_factory=lambda: {"all_of": [], "any_of": [], "none_of": []}
    )


class BlendRequest(BaseModel):
    blocks: list[BlockInput]
    seed: int = 42
    min_confidence: float = Field(default=DEFAULT_MIN_CONFIDENCE, ge=0.5, le=1.0)
    max_dark_ratio: float = Field(default=DEFAULT_MAX_DARK_RATIO, ge=0.5, le=1.0)


class ImageResult(BaseModel):
    image_sk: int
    nasa_id: str
    selection_rate: float
    selection_count: int
    shown_count: int


class BallotImage(BaseModel):
    image_sk: int
    nasa_id: str
    utility: float
    attribute_match: float
    base_appeal: float
    was_selected: bool
    matched_attrs: list[str]


class SampleBallot(BaseModel):
    voter_number: int
    ballot_number: int
    block_id: str
    images: list[BallotImage]


class BlockResult(BaseModel):
    block_id: str
    label: str
    voter_count: int
    ballot_count: int
    top_13: list[ImageResult]
    sample_ballots: list[SampleBallot]


class BlendResponse(BaseModel):
    seed: int
    total_voters: int
    total_ballots: int
    eligible_images: int
    min_confidence: float
    max_dark_ratio: float
    blocks: list[BlockResult]
    blended: list[ImageResult]
    overlap: dict[str, list[int]]


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------


@router.get("/presets")
def get_presets(conn: duckdb.DuckDBPyConnection = Depends(get_db)):
    """Return preset block configs and available attribute codes."""
    try:
        rows = conn.execute(
            "SELECT DISTINCT attribute_code FROM feature_image_attribute "
            "WHERE is_accepted = true ORDER BY attribute_code"
        ).fetchall()
        attributes = [r[0] for r in rows]
    except Exception:
        attributes = []

    return {"blocks": PRESET_BLOCKS, "attributes": attributes}


@router.post("/simulate", response_model=BlendResponse)
def simulate(req: BlendRequest, conn: duckdb.DuckDBPyConnection = Depends(get_db)):
    """Run in-memory voting simulation and return per-block + blended top-13."""
    if not req.blocks:
        raise HTTPException(400, "At least one block required")

    active_blocks = [b for b in req.blocks if b.voter_count > 0]
    if not active_blocks:
        raise HTTPException(400, "At least one block must have voter_count > 0")

    # Load filtered images: no duplicates, no near-black, higher confidence
    image_attrs, sid_map = _load_simulation_images(
        conn,
        min_confidence=req.min_confidence,
        max_dark_ratio=req.max_dark_ratio,
    )
    all_image_sks = list(image_attrs.keys())

    if not all_image_sks:
        raise HTTPException(500, "No eligible images found after filtering")

    # Generate deterministic base appeal per image (same as block_generator)
    base_appeals = {}
    for sk in all_image_sks:
        h = hashlib.sha256(f"appeal:{sk}:{req.seed}".encode()).hexdigest()
        base_appeals[sk] = int(h[:8], 16) / 0xFFFFFFFF

    # Simulate voting per block
    total_voters = 0
    total_ballots = 0
    blended_selected: dict[int, int] = defaultdict(int)
    blended_shown: dict[int, int] = defaultdict(int)
    block_results: list[BlockResult] = []

    for block_input in active_blocks:
        block_config = VotingBlockConfig(
            block_id=block_input.block_id,
            label=block_input.label or block_input.block_id,
            voter_count=block_input.voter_count,
            votes_per_voter=block_input.votes_per_voter,
            preference_rules=PreferenceRule(
                all_of=block_input.preference_rules.get("all_of", []),
                any_of=block_input.preference_rules.get("any_of", []),
                none_of=block_input.preference_rules.get("none_of", []),
            ),
            preference_weight=block_input.preference_weight,
            randomness_weight=block_input.randomness_weight,
        )

        rng = random.Random(req.seed + hash(block_input.block_id))
        selected_counts: dict[int, int] = defaultdict(int)
        shown_counts: dict[int, int] = defaultdict(int)
        ballot_count = 0
        sample_ballots: list[SampleBallot] = []

        # Decide which ballots to capture as samples.
        # Capture the 1st ballot of the 1st, middle, and last voter.
        sample_voter_indices = set()
        if block_input.voter_count >= 3:
            sample_voter_indices = {0, block_input.voter_count // 2, block_input.voter_count - 1}
        elif block_input.voter_count > 0:
            sample_voter_indices = {0}

        for voter_i in range(block_input.voter_count):
            for vote_i in range(block_input.votes_per_voter):
                shown = rng.sample(
                    all_image_sks, min(BATCH_SHOW_COUNT, len(all_image_sks))
                )
                scored = []
                for img_sk in shown:
                    attrs = image_attrs.get(img_sk, set())
                    u = compute_block_utility(
                        attrs, block_config, base_appeals[img_sk], rng
                    )
                    scored.append((img_sk, u))

                scored.sort(key=lambda x: x[1], reverse=True)
                winners = {s[0] for s in scored[:BATCH_SELECT_COUNT]}

                # Capture sample ballot (1st ballot of selected voters)
                if voter_i in sample_voter_indices and vote_i == 0:
                    # Build the preference rules' attribute set for matching display
                    rule_attrs = set(
                        block_input.preference_rules.get("all_of", [])
                        + block_input.preference_rules.get("any_of", [])
                    )

                    ballot_images = []
                    for img_sk, utility in scored:
                        attrs = image_attrs.get(img_sk, set())
                        match_score = compute_attribute_match(attrs, block_config)
                        matched = sorted(attrs & rule_attrs) if rule_attrs else []
                        ballot_images.append(
                            BallotImage(
                                image_sk=img_sk,
                                nasa_id=sid_map.get(img_sk, ""),
                                utility=round(utility, 4),
                                attribute_match=round(match_score, 4),
                                base_appeal=round(base_appeals[img_sk], 4),
                                was_selected=img_sk in winners,
                                matched_attrs=matched,
                            )
                        )

                    sample_ballots.append(
                        SampleBallot(
                            voter_number=voter_i + 1,
                            ballot_number=vote_i + 1,
                            block_id=block_input.block_id,
                            images=ballot_images,
                        )
                    )

                for img_sk, _ in scored:
                    shown_counts[img_sk] += 1
                    blended_shown[img_sk] += 1
                    if img_sk in winners:
                        selected_counts[img_sk] += 1
                        blended_selected[img_sk] += 1

                ballot_count += 1

        total_voters += block_input.voter_count
        total_ballots += ballot_count

        # Compute selection rates and pick top-13
        rates = {
            sk: selected_counts[sk] / max(shown_counts[sk], 1)
            for sk in all_image_sks
            if shown_counts[sk] > 0
        }
        top_sks = sorted(rates, key=rates.get, reverse=True)[:13]

        block_results.append(
            BlockResult(
                block_id=block_input.block_id,
                label=block_input.label or block_input.block_id,
                voter_count=block_input.voter_count,
                ballot_count=ballot_count,
                top_13=[
                    ImageResult(
                        image_sk=sk,
                        nasa_id=sid_map.get(sk, ""),
                        selection_rate=rates.get(sk, 0.0),
                        selection_count=selected_counts[sk],
                        shown_count=shown_counts[sk],
                    )
                    for sk in top_sks
                ],
                sample_ballots=sample_ballots,
            )
        )

    # Blended top-13
    blended_rates = {
        sk: blended_selected[sk] / max(blended_shown[sk], 1)
        for sk in all_image_sks
        if blended_shown[sk] > 0
    }
    blended_top_sks = sorted(blended_rates, key=blended_rates.get, reverse=True)[:13]
    blended_top_set = set(blended_top_sks)

    blended_results = [
        ImageResult(
            image_sk=sk,
            nasa_id=sid_map.get(sk, ""),
            selection_rate=blended_rates.get(sk, 0.0),
            selection_count=blended_selected[sk],
            shown_count=blended_shown[sk],
        )
        for sk in blended_top_sks
    ]

    # Compute overlap: which block top-13 images also appear in blended top-13
    overlap: dict[str, list[int]] = {}
    for br in block_results:
        block_top_set = {img.image_sk for img in br.top_13}
        shared = sorted(block_top_set & blended_top_set)
        overlap[br.block_id] = shared

    return BlendResponse(
        seed=req.seed,
        total_voters=total_voters,
        total_ballots=total_ballots,
        eligible_images=len(all_image_sks),
        min_confidence=req.min_confidence,
        max_dark_ratio=req.max_dark_ratio,
        blocks=block_results,
        blended=blended_results,
        overlap=overlap,
    )
